packages/wdio-browserstack-service/src/insights-handler.ts

```diff
@@ -0,0 +1,457 @@
+    private _platformMeta?: PlatformMeta
+    private _commands: Record<string, BeforeCommandArgs & AfterCommandArgs> = {}
+    private _gitConfigPath?: string
+    private _requestQueueHandler = RequestQueueHandler.getInstance()
+
+    constructor (private _browser: Browser<'async'> | MultiRemoteBrowser<'async'>, browserCaps?: Capabilities.Capabilities, isAppAutomate?: boolean, sessionId?: string, private _framework?: string) {
+        this._requestQueueHandler.start()
+
+        /* istanbul ignore next */
+        this._platformMeta = {
+            browserName: browserCaps?.browserName,
+            browserVersion: browserCaps?.browserVersion,
+            platformName: browserCaps?.platformName,
+            caps: browserCaps,
+            sessionId: sessionId,
+            product: isAppAutomate ? 'app-automate' : 'automate'
+        }
+    }
+
+    async before () {
+        if (isBrowserstackSession(this._browser)) {
+            await this._browser.execute(`browserstack_executor: {"action": "annotate", "arguments": {"data": "ObservabilitySync:${Date.now()}","level": "debug"}}`)
+        }
+
+        const gitMeta = await getGitMetaData()
+        if (gitMeta) {
+            this._gitConfigPath = gitMeta['root']
+        }
+    }
+
+    async beforeHook (test: Frameworks.Test, context: any) {
+        const fullTitle = `${test.parent} - ${test.title}`
+        const hookId = uuidv4()
+        this._tests[fullTitle] = {
+            uuid: hookId,
+            startedAt: (new Date()).toISOString()
+        }
+        this.attachHookData(context, hookId)
+        if (this._framework == 'mocha') await this.sendTestRunEvent(test, 'HookRunStarted')
+    }
+
+    async afterHook (test: Frameworks.Test, result: Frameworks.TestResult) {
+        const fullTitle = getUniqueIdentifier(test)
+        if (this._tests[fullTitle]) {
+            this._tests[fullTitle].finishedAt = (new Date()).toISOString()
+        } else {
+            this._tests[fullTitle] = {
+                finishedAt: (new Date()).toISOString()
+            }
+        }
+        if (this._framework == 'mocha') await this.sendTestRunEvent(test, 'HookRunFinished', result)
+    }
+
+    async beforeTest (test: Frameworks.Test) {
+        const fullTitle = getUniqueIdentifier(test)
+        this._tests[fullTitle] = {
+            uuid: uuidv4(),
+            startedAt: (new Date()).toISOString()
+        }
+        await this.sendTestRunEvent(test, 'TestRunStarted')
+    }
+
+    async afterTest (test: Frameworks.Test, result: Frameworks.TestResult) {
+        const fullTitle = getUniqueIdentifier(test)
+        this._tests[fullTitle] = {
+            ...(this._tests[fullTitle] || {}),
+            finishedAt: (new Date()).toISOString()
+        }
+        await this.sendTestRunEvent(test, 'TestRunFinished', result)
+    }
+
+    /**
+      * Cucumber Only
+      */
+
+    async beforeScenario (world: ITestCaseHookParameter) {
+        let pickleData = world.pickle
```

Comment:
I personally would not create variables for these kind of things given there doesn't seem to be a need for it but I don't want to be too picky here. However if we create variables we don't modify we should at least define them as `const`
```suggestion
        const pickleData = world.pickle
```